### DISCUSSION TITLE
Feature Request: Syntax Sugar for AddRange in List Initialization

### DISCUSSION THREAD

**Discussion by Dogwei** [dotnet/csharplang#7790] [Language Ideas]
I would like to propose a new syntax for Roslyn. Here is the idea:

Currently, if we want to initialize a list with an array and some other elements, we have to do something like this:

```csharp
var array = new int[] { 1, 2, 3 };
var list = new List<int>();
list.Add(1);
list.AddRange(array);
list.Add(2);
list.AddRange(array);
list.Add(3);
```

Referring to the syntax of other languages, I propose a new syntax that allows us to do the same thing more concisely:

```csharp
var array = new int[] { 1, 2, 3 };
var list = new List<int>
{
    1,
    ...array,
    2,
    ...array,
    3,
};
```

In this proposed syntax, `...array` would be equivalent to `AddRange(array)`. This would make the code more readable and easier to write.

**Comment by kyoyama-kazusa** [comment]
Ah.

Here is the solution: C# 12 introduces the syntax "Collection Expression" to describe such usages:

```csharp
var list = (List<int>)[1, .. array, 2, .. array, 3];

// Or
List<int> list = [1, .. array, 2, .. array, 3];
```

Because a collection expression doesn't contain a natural type, we cannot use `var` here. In C# 13, we may know the natural type of such expression. Therefore, we can write

```csharp
// Here 'list' may be a 'List<int>'.
var list = [1, .. array, 2, .. array, 3];
```
